Stripe/STPCustomerContext.m

```diff
@@ -0,0 +1,151 @@
+#import "STPWeakStrongMacros.h"
+
+static NSTimeInterval const DefaultCachedCustomerMaxAge = 60;
+
+@interface STPCustomerContext ()
+
+@property (nonatomic) STPAPIClient *apiClient;
+@property (nonatomic) STPCustomer *customer;
+@property (nonatomic) NSDate *customerRetrievedDate;
+@property (nonatomic) STPEphemeralKeyManager *keyManager;
+
+@end
+
+@implementation STPCustomerContext
+
++ (instancetype)sharedInstance {
+    static STPCustomerContext *sharedInstance;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [self new];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)initWithKeyProvider:(nonnull id<STPEphemeralKeyProvider>)keyProvider {
+    STPEphemeralKeyManager *keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider
+                                                                                  apiVersion:[STPAPIClient apiVersion]];
+    return [self initWithKeyManager:keyManager];
+}
+
+- (instancetype)initWithKeyManager:(STPEphemeralKeyManager *)keyManager {
+    self = [super init];
+    if (self) {
+        _cachedCustomerMaxAge = DefaultCachedCustomerMaxAge;
+        _keyManager = keyManager;
+        [self retrieveCustomer:nil];
+    }
+    return self;
+}
+
+- (void)setKeyProvider:(id<STPEphemeralKeyProvider>)keyProvider {
+    _keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider
+                                                           apiVersion:[STPAPIClient apiVersion]];
+    [self retrieveCustomer:nil];
+}
+
+- (void)setCustomer:(STPCustomer *)customer {
+    _customer = customer;
+    _customerRetrievedDate = [NSDate date];
+}
+
+- (BOOL)shouldUseCachedCustomer {
+    if (!self.customer || !self.customerRetrievedDate) {
+        return NO;
+    }
+    NSDate *now = [NSDate date];
+    return [now timeIntervalSinceDate:self.customerRetrievedDate] < self.cachedCustomerMaxAge;
+}
+
+- (void)retrieveCustomer:(STPCustomerCompletionBlock)completion {
+    if ([self shouldUseCachedCustomer]) {
+        if (completion) {
+            completion(self.customer, nil);
+        }
+        return;
+    }
+    if (!self.keyManager) {
+        completion(nil, [NSError stp_customerContextMissingKeyProviderError]);
+        return;
+    }
+    [self.keyManager getCustomerKey:^(STPEphemeralKey *ephemeralKey, NSError *retrieveKeyError) {
+        if (retrieveKeyError) {
+            if (completion) {
+                completion(nil, retrieveKeyError);
+            }
+            return;
+        }
+        [STPAPIClient retrieveCustomerUsingKey:ephemeralKey completion:^(STPCustomer *customer, NSError *error) {
+            if (customer) {
+                self.customer = customer;
+            }
+            if (completion) {
+                completion(customer, error);
+            }
+        }];
+    }];
+}
+
+- (void)attachSourceToCustomer:(id<STPSourceProtocol>)source completion:(STPErrorBlock)completion {
```

Comment:
This changes the sources array on customer, but doesn't update the cached customer object or clear out the cache setting so your next fetch of the customer may be out of sync. Should force a re-fetch after this and/or add some way to force update customers ignoring the cache.